sqlite3_cache/cache.py

```diff
@@ -522,3 +525,99 @@ def ttl_many(self, keys: List[str]) -> Dict[str, int]:
+            self._con.commit()
+
+        return results
+
+    def get_all_keys(self) -> List[str]:
+        """
+        Get all keys that exist in the cache for currently valid cache items.
+
+        :return: List of cache keys in sort order.
+        """
+        fetched: List[Tuple[str, Any]] = self._con.execute(self._get_keys_sql).fetchall()
+
+        if not fetched:
+            return []
+
+        return self._filter_key_result_list(fetched)
+
+    def find_keys_starting_with(self, pattern: str) -> List[str]:
+        """
+        Find keys that start with the given pattern.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+        Will only return keys that exist in the cache for currently valid cache items.
+
+        :param pattern: The pattern to match at the start of the key.
+        :return: List of matching cache keys in sort order.
+        """
+        data = {"pattern": f"{''}{pattern}%"}
+        fetched: List[Tuple[str, Any]] = self._con.execute(self._find_matching_keys_sql, data).fetchall()
+
+        if not fetched:
+            return []
+
+        return self._filter_key_result_list(fetched)
+
+    def find_matching_keys(self, pattern: str) -> List[str]:
+        """
+        Find keys that contain the given pattern anywhere in the string.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+        Will only return keys that exist in the cache for currently valid cache items.
+
+        :param pattern: The pattern to find in matching keys.
+        :return: List of matching cache keys in sort order.
+        """
+        data = {"pattern": f"{'%'}{pattern}%"}
+        fetched: List[Tuple[str, Any]] = self._con.execute(self._find_matching_keys_sql, data).fetchall()
+
+        if not fetched:
+            return []
+
+        return self._filter_key_result_list(fetched)
+
+    def clear_keys_starting_with(self, pattern: str) -> None:
+        """
+        Clear all keys from the cache that start with the given pattern.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+
+        :param pattern: The pattern to match at the start of the key.
+        """
+        data = {"pattern": f"{''}{pattern}%"}
+        self._con.execute(self._clear_starting_with_sql, data)
+        self._con.commit()
+
+    def clear_matching_keys(self, pattern: str) -> None:
+        """
+        Clear all keys from the cache that contain the given pattern anywhere in the string.
+        Matching follows the SQLite specification for the LIKE operator, so
+        it will match 'A' to 'a', but not 'Ä' to 'ä'.
+
+        :param pattern: The pattern to find in matching keys.
+        """
+        data = {"pattern": f"{'%'}{pattern}%"}
```

Comment:
Pattern should probably be `f"%{pattern}%"`. Alternatiely, you could leave the `%`'s out and make this a more generic method, and refactor the "starting_with" method to use it.

Reply:
it was hard for me to find a way to do this that both felt like the simplest clearest API to declare and also passed linting.

Reply:
Fixed

Reply:
I think you could do something like this:

```python
def find_matching_keys(self, pattern: str) -> List[str]:
    # Any custom pattern can be used here
    data = {"pattern": pattern}
    fetched: List[Tuple[str, Any]] = self._con.execute(self._find_matching_keys_sql, data).fetchall()
    if not fetched:
        return []

    return self._filter_key_result_list(fetched)

def find_keys_keys_starting_with(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"{pattern}%")

def find_keys_keys_ending_with(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"%{pattern}")

def find_keys_keys_containing(self, pattern: str) -> List[str]:
    return self.find_matching_keys(f"%{pattern}%")
```

Reply:
I didn't think you'd want to expose the SQL `%` syntax of the `LIKE` operator in the API itself.  But now that I know it's OK with you, it will simplify things!

Reply:
OK, I took your suggestion and rewrote the methods this way, including for the `clear_keys` methods, and updated `docs/ref.md` as well.